src/Proxy.NET/Program.cs

```diff
@@ -0,0 +1,177 @@
+                        Auth.Type.None => null,
+                        _ => throw new ArgumentException("Mismatched auth type or HTTP verb")
+                    };
+
+                    await next.Invoke();
+                }
+                catch (JsonException ex)
+                {
+                    logger.LogInformation(ex, "JSON exception.");
+                    await WriteOpenAIExceptionFormat(context, ex.Message, (int)HttpStatusCode.BadRequest);
+                }
+                catch (ArgumentException ex)
+                {
+                    logger.LogWarning(ex, "Argument exception.");
+                    await WriteOpenAIExceptionFormat(context, ex.Message, (int)HttpStatusCode.BadRequest);
+                }
+                catch (HttpRequestException ex)
+                {
+                    logger.LogInformation(ex, "An error occurred while processing the request.");
+                    await WriteOpenAIExceptionFormat(context, ex.Message, (int?)ex.StatusCode ?? (int)HttpStatusCode.ServiceUnavailable);
+                }
+                catch (NpgsqlException ex)
+                {
+                    logger.LogError(ex, "Database error.");
+                    await WriteOpenAIExceptionFormat(context, ex.Message, (int)HttpStatusCode.ServiceUnavailable);
+                }
+                catch (TaskCanceledException ex)
+                {
+                    logger.LogWarning(ex, "The request was canceled.");
+                    await WriteOpenAIExceptionFormat(context, ex.Message, (int)HttpStatusCode.GatewayTimeout);
+                }
+                catch (BadHttpRequestException ex)
+                {
+                    logger.LogWarning(ex, "The request was bad.");
+                    await WriteOpenAIExceptionFormat(context, ex.Message, (int)HttpStatusCode.BadRequest);
+                }
+                catch (Exception ex)
+                {
+                    logger.LogError(ex, "An unexpected error occurred.");
+                    await WriteOpenAIExceptionFormat(context, ex.Message, (int)HttpStatusCode.InternalServerError);
+                }
+            }
+        );
+
+        app.Run();
+    }
+
+    /// <summary>
+    /// Writes the OpenAI exception format to the HTTP response.
+    /// </summary>
+    /// <param name="context">The HttpContext.</param>
+    /// <param name="msg">The error message.</param>
+    /// <param name="statusCode">The status code.</param>
+    /// <returns>A task representing the asynchronous operation.</returns>
+    private static async Task WriteOpenAIExceptionFormat(HttpContext context, string msg, int statusCode)
+    {
+        var jsonMsg = new { error = new { code = statusCode, message = msg } };
+        context.Response.StatusCode = statusCode;
+        context.Response.ContentType = "application/json";
+        await context.Response.WriteAsJsonAsync(jsonMsg);
+    }
```

Comment:
We should be using the `Results` API to return here, as that takes care of the serialization and everything for us.

So this method would probably be removed, and we'd create a `record`, as `record OpenAIError(int code, string message)` and then handle the serialization a bit better. This means we could then do `Results.BadRequest(new OpenAIError(400, "some message"))` rather than having to do our own serialization and response writing.

Reply:
I went with var error = new OpenAIError(statusCode, msg);
        await Results.Json(new { error }, statusCode: statusCode).ExecuteAsync(context);

as need to return the status code of the exception - eg 429 for rate limit exceeded rather than always returning 400 bad request